src/basic_hierarchy/common/HierarchyBuilder.java

```diff
@@ -0,0 +1,395 @@
+				if ( candidateGroup != null ) {
+					nearestParent = candidateGroup;
+					nearestParentHeight = getGroupBranchIds( candidateGroup ).length;
+				}
+
+				// Try to find nearest parent in artificial groups
+				candidateGroup = findNearestAncestor( groups, groupBranchIds, nearestParentHeight );
+				if ( candidateGroup != null ) {
+					nearestParent = candidateGroup;
+					nearestParentHeight = getGroupBranchIds( candidateGroup ).length;
+				}
+
+				if ( nearestParent != null ) {
+					BasicNode newParent = nearestParent;
+
+					for ( int j = nearestParentHeight; j < groupBranchIds.length - 1; ++j ) {
+						buf.setLength( 0 );
+
+						String newGroupId = buf
+							.append( newParent.getId() )
+							.append( Constants.HIERARCHY_BRANCH_SEPARATOR )
+							.append( groupBranchIds[j] )
+							.toString();
+
+						// Add an empty group
+						BasicNode newGroup = new BasicNode(
+							newGroupId, newParent, useSubtree
+						);
+
+						// Create proper parent-child relations
+						newParent.addChild( newGroup );
+						newGroup.setParent( newParent );
+
+						artificialGroups.add( newGroup );
+
+						newParent = newGroup;
+					}
+
+					// Add missing links
+					newParent.addChild( group );
+					group.setParent( newParent );
+				}
+				else {
+					throw new RuntimeException(
+						String.format(
+							"Could not find nearest parent for '%s'. This means that something went seriously wrong.",
+							group.getId()
+						)
+					);
+				}
+			}
+		}
+
+		List<BasicNode> allGroups = new ArrayList<BasicNode>( artificialGroups );
+		allGroups.addAll( groups );
+
+		return allGroups;
+	}
+
+	/**
+	 * Fixes gaps in breadth (missing siblings) by creating empty groups where needed.
+	 * <p>
+	 * Such gaps appear when the source file did not list these groups (since they were empty),
+	 * but their existence can be inferred from IDs of existing groups.
+	 * </p>
+	 * 
+	 * @param root
+	 *            the root group
+	 * @param groups
+	 *            the original collection of groups
+	 * @param useSubtree
+	 *            whether the cntroid calculation should also include child groups' instances.
+	 * @return the complete 'fixed' collection of groups, filled with artificial groups
+	 */
+	private static List<BasicNode> fixBreadthGaps( BasicNode root, List<BasicNode> groups, boolean useSubtree )
+	{
+		List<BasicNode> artificialGroups = new ArrayList<BasicNode>();
+
+		Comparator<Node> groupComparator = new NodeIdComparator();
+		StringBuilder buf = new StringBuilder();
+
+		Queue<BasicNode> pendingGroups = new LinkedList<BasicNode>();
+		pendingGroups.add( root );
+
+		while ( !pendingGroups.isEmpty() ) {
+			BasicNode currentGroup = pendingGroups.remove();
+
+			List<Node> children = currentGroup.getChildren();
+			List<Node> newChildren = new LinkedList<Node>();
+
+			for ( int i = 0; i < children.size(); ++i ) {
+				Node childGroup = children.get( i );
```

Comment:
Where do we assure that children will have null values when there will be no child node?